Purchases/Caching/RCDeviceCache.m

```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
             *groupedAttributes = self.storedAttributesForAllUsers.mutableCopy;
         [groupedAttributes removeObjectForKey:appUserID];
-        [self.userDefaults setObject:groupedAttributes forKey:RCSubscriberAttributesKey];
+        [self storeAttributesForAllUsers:groupedAttributes];
     }
 }
 
 # pragma mark - subscriber attributes migration from per-user key to grouped key
 
-- (void)migrateSubscriberAttributesIfNeededForAppUserID:(nonnull NSString *)appUserID {
+- (void)cleanupSubscriberAttributes {
     @synchronized (self) {
-        NSDictionary *legacySubscriberAttributes = [self valueForLegacySubscriberAttributes:appUserID];
-        if (legacySubscriberAttributes != nil) {
-            [self migrateSubscriberAttributes:legacySubscriberAttributes withAppUserID:appUserID];
-        }
+        [self migrateSubscriberAttributes];
+        [self deleteSyncedSubscriberAttributesForOtherUsers];
     }
 }
 
-- (void)migrateSubscriberAttributes:(nonnull NSDictionary *)legacyAttributes
-                      withAppUserID:(nonnull NSString *)appUserID {
-    NSMutableDictionary *allSubscriberAttributes = ([self.userDefaults dictionaryForKey:RCSubscriberAttributesKey]
-                                                    ?: @{}).mutableCopy;
-    NSMutableDictionary *currentAttributesForAppUserID = allSubscriberAttributes[appUserID]
-                                                         ?: @{}.mutableCopy;
-    NSMutableDictionary *mutableLegacyAttributes = legacyAttributes.mutableCopy;
-    [mutableLegacyAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+- (void)migrateSubscriberAttributes {
+    NSArray *appUserIDsWithLegacyAttributes = [self appUserIDsWithLegacyAttributes];
+    NSMutableDictionary *attributesInNewFormat = self.storedAttributesForAllUsers.mutableCopy;
+    for (NSString *appUserID in appUserIDsWithLegacyAttributes) {
+        NSDictionary *legacyAttributes = ([self valueForLegacySubscriberAttributes:appUserID] ?: @{}).mutableCopy;
+        NSMutableDictionary *currentAttributesForAppUserID = (legacyAttributes ?: @{}).mutableCopy;
+        NSMutableDictionary *allAttributesForUser = legacyAttributes.mutableCopy;
+        [allAttributesForUser addEntriesFromDictionary:currentAttributesForAppUserID];
+
+        attributesInNewFormat[appUserID] = allAttributesForUser;
+
+        NSString *legacyAttributesKey = [self legacySubscriberAttributesCacheKeyForAppUserID:appUserID];
+        [self.userDefaults removeObjectForKey:legacyAttributesKey];
+    }
+    [self storeAttributesForAllUsers:attributesInNewFormat];
+}
+
+- (void)deleteSyncedSubscriberAttributesForOtherUsers {
+    NSDictionary<NSString *, NSDictionary *> *allStoredAttributes = self.storedAttributesForAllUsers;
+    NSMutableDictionary<NSString *, NSDictionary *> *filteredAttributes = [[NSMutableDictionary alloc] init];
+
+    NSString *currentAppUserID = self.cachedAppUserID;
+    NSParameterAssert(currentAppUserID);
```

Comment:
Is this to avoid a crash when `NSUserDefaults` is cleared?

Reply:
it's for when `NSUserDefaults` is cleared, yes, but it's not really meant to prevent a crash - it'll crash anyway. Just slightly easier to debug if this happens. 
 it's a fairly common practice in Obj-c to `NSParamAssert` stuff that you want to guarantee is non-null, since no-ops and nullability issues can be annoying as hell to debug.

Reply:
oh I see, got it! But the other PR that added the observer on NSUserDefaults will still work right?

Reply:
yep. so the only way to reach this line with a nil appUserID _should_ be if you never had one in the first place.

Reply:
I noticed this is calling `self.cachedAppUserID` instead of passing it to `deleteSyncedSubscriberAttributesForOtherUsers` from the `RCIdentityManager`. Do you think we could have some sort of race condition if configure gets called twice? 🤔

Reply:
that's a valid concern. 
however: this method is private, and only called within a synchronized block, so passing in the appUserID would have the same effect as reading it from disk, right? 
also, the configure method actually creates a new instance of purchases and its dependencies (including this one), so their variables are independent

Reply:
merging for now, but let's keep the convo open